src/optimizer/query_to_operator_transformer.cpp

```diff
@@ -246,6 +250,67 @@ void QueryToOperatorTransformer::Visit(parser::InsertStatement *op) {
     insert_expr->PushChild(output_expr_);
     output_expr_ = insert_expr;
   } else {
+    /*auto storage_manager = storage::StorageManager::GetInstance();
+    auto database =
+        storage_manager->GetDatabaseWithOid(target_table->GetDatabaseOid());
+    auto table = database->GetTableWithOid(target_table->GetTableOid());
+
+    auto *schema = table->GetSchema();
+    */
+
+    // column_objects represents the columns for the current table as defined in
+    // its schema
+    auto column_objects = target_table->GetColumnObjects();
+    // INSERT INTO table_name VALUES (val1, val2, ...), (val_a, val_b, ...), ...
+    if (op->columns.empty()) {
+      for (uint32_t tuple_idx = 0; tuple_idx < op->insert_values.size();
+           tuple_idx++) {
+        auto &values = (op->insert_values)[tuple_idx];
+        if (values.size() > column_objects.size()) {
+          throw CatalogException(
+              "ERROR:  INSERT has more expressions than target columns");
+        } else if (values.size() < column_objects.size()) {
+          for (oid_t i = values.size(); i != column_objects.size(); ++i) {  // check whether null values are allowed in the rest of the columns
+            if (column_objects[i]->IsNotNull()) {
```

Comment:
Here we also need to check whether a default value is available for the current column. How do I do that? The `ColumnCatalogObject` instances don't have any method for that.